### PR TITLE
Add interface for resource usage listener to spi.

### PR DESCRIPTION
**Pull request by toregge**
Propagate resource usage to service layer.

@geirst : please review
@baldersheim : FYI

Supercedes #16041

### CODE REVIEW

File: searchcore/src/vespa/searchcore/proton/persistenceengine/resource_usage_tracker.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      _listener(nullptr),
+      _disk_mem_usage_notifier(disk_mem_usage_notifier)
+{
+    _disk_mem_usage_notifier.addDiskMemUsageListener(this);
+}
+
+ResourceUsageTracker::~ResourceUsageTracker()
+{
+    _disk_mem_usage_notifier.removeDiskMemUsageListener(this);
+    std::lock_guard guard(_lock);
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/persistenceengine/resource_usage_tracker.cpp:49]: Why are we taking the lock here?

**Review by geirst — APPROVED** [review]
👍

File: searchcore/src/vespa/searchcore/proton/persistenceengine/resource_usage_tracker.cpp (hunk 1 above)
Comment (toregge) [reply searchcore/src/vespa/searchcore/proton/persistenceengine/resource_usage_tracker.cpp]: I've removed the guard.

File: persistence/src/vespa/persistence/spi/persistenceprovider.h [hunk 2]
```diff
@@ -374,6 +377,12 @@ struct PersistenceProvider
      * source1 and source2 should be stored in the target bucket.
      */
     virtual Result join(const Bucket& source1, const Bucket& source2, const Bucket& target, Context&) = 0;
+
+    /*
+     * Register a listener for updates to resource usage.
+     */
+    virtual std::unique_ptr<vespalib::IDestructorCallback> register_resource_usage_listener(IResourceUsageListener& listener) = 0;
```
[the remark this answers is not in the document]
Comment (toregge) [reply persistence/src/vespa/persistence/spi/persistenceprovider.h:385]: Description of semantics of return type has been added.

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**